tests/atoms/base/ClassServerUTest.cxxtest

```diff
@@ -0,0 +1,236 @@
+
+		TS_ASSERT(!class_server->getFactory(NODE_A));
+	}
+
+	void test_replace_descendants_factory()
+	{
+		class_server->addFactory(NODE_A, factory_a);
+
+		class_server->addFactory(NODE_B, factory_b);
+
+		TS_ASSERT_EQUALS(factory_b, class_server->getFactory(NODE_C));
+	}
+
+	void test_factory_from_other_hierarchy_is_not_replaced()
+	{
+		class_server->addFactory(NODE_A, factory_a);
+
+		class_server->addFactory(NODE_B, factory_b);
+
+		TS_ASSERT_EQUALS(factory_a, class_server->getFactory(NODE_D));
+	}
+
+	void test_not_parent_factory_is_not_replaced()
+	{
+		class_server->addFactory(NODE_C, factory_c);
+
+		class_server->addFactory(NODE_B, factory_b);
+
+		TS_ASSERT_EQUALS(factory_c, class_server->getFactory(NODE_C));
+	}
+
+	void test_add_validator()
+	{
+		class_server->addValidator(NODE_B, validator_b);
+
+		TS_ASSERT_EQUALS(validator_b, class_server->getValidator(NODE_B));
+	}
+
+	void test_replace_descendants_validators()
+	{
+		class_server->addValidator(NODE_A, validator_a);
+
+		TS_ASSERT_EQUALS(validator_a, class_server->getValidator(NODE_A));
+		TS_ASSERT_EQUALS(validator_a, class_server->getValidator(NODE_B));
+		TS_ASSERT_EQUALS(validator_a, class_server->getValidator(NODE_C));
+	}
+
+	void test_parent_validator_is_not_replaced()
+	{
+		class_server->addValidator(NODE_B, validator_b);
+
+		TS_ASSERT(!class_server->getValidator(NODE_A));
+	}
+
+	void test_replace_descendants_validator()
+	{
+		class_server->addValidator(NODE_A, validator_a);
+
+		class_server->addValidator(NODE_B, validator_b);
+
+		TS_ASSERT_EQUALS(validator_b, class_server->getValidator(NODE_C));
+	}
+
+	void test_validator_from_other_hierarchy_is_not_replaced()
+	{
+		class_server->addValidator(NODE_A, validator_a);
+
+		class_server->addValidator(NODE_B, validator_b);
+
+		TS_ASSERT_EQUALS(validator_a, class_server->getValidator(NODE_D));
+	}
+
+	void test_explicitly_set_validator_is_not_replaced()
+	{
+		TS_SKIP("fails");
+		class_server->addValidator(NODE_C, validator_c);
+
+		class_server->addValidator(NODE_B, validator_b);
+
+		TS_ASSERT_EQUALS(validator_c, class_server->getValidator(NODE_C));
+	}
+
+	void test_add_validating_factory()
+	{
+		class_server->addValidator(NODE_B, validator_b);
+
+		TS_ASSERT(class_server->getFactory(NODE_B));
+	}
+
+	void test_add_validating_factory_to_descendants()
+	{
+		class_server->addValidator(NODE_B, validator_b);
+
+		TS_ASSERT(class_server->getFactory(NODE_C));
+	}
+
+	void test_explicitly_set_factory_is_not_replaced_by_validating_factory()
+	{
+		class_server->addFactory(NODE_B, factory_b);
+
+		class_server->addValidator(NODE_B, validator_b);
+
+		TS_ASSERT_EQUALS(factory_b, class_server->getFactory(NODE_B));
+	}
+
+	void test_implicitly_set_factory_is_not_replaced_by_validating_factory()
+	{
+		TS_SKIP("fails");
+		class_server->addFactory(NODE_A, factory_a);
+
+		class_server->addValidator(NODE_B, validator_b);
+
+		TS_ASSERT_EQUALS(factory_b, class_server->getFactory(NODE_B));
```

Comment:
I think you want instead
```
TS_ASSERT_EQUALS(validator_b, class_server->getFactory(NODE_B));
```

Reply:
Thank you for catching it!
I was going to check: `TS_ASSERT_EQUALS(factory_a, class_server->getFactory(NODE_B));`
Raised PR https://github.com/opencog/atomspace/pull/2028 with use case description